[SQL] Rust compilation errors

## Bug report: `group-by_050`

     ### Spark SQL

     **Schema**

     _empty_

     **Query**

     ```sql
     CREATE VIEW group-by_050 AS
     SELECT count(*)
     FROM VALUES (named_struct('b', array(map(1, 2, 2, 3), map(1, 3)))), (named_struct('b', array(map(2, 3), map(1, 3)))), (named_struct('b', array(map(2, 3, 1, 2), map(1, 3)))) as t(a)
     GROUP BY a;
     ```

     **Expected result**

     ```
     2
     1
     ```

     ### Feldera SQL (translated)

     **Schema**

     _empty_

     **Query**

     ```sql
     CREATE VIEW "group-by_050" AS
     SELECT COUNT(*)
     FROM (
       VALUES
         (CAST(ROW(ARRAY[MAP[1, 2, 2, 3], MAP[1, 3]]) AS ROW(b MAP<INT, INT> ARRAY))),
         (CAST(ROW(ARRAY[MAP[2, 3], MAP[1, 3]]) AS ROW(b MAP<INT, INT> ARRAY))),
         (CAST(ROW(ARRAY[MAP[2, 3, 1, 2], MAP[1, 3]]) AS ROW(b MAP<INT, INT> ARRAY)))
     ) AS t(a)
     GROUP BY a
     ```

     ### Status

     | Phase | Status |
     |---|---|
     | Translation | `success` |
     | Compilation | `ok` |
     | Execution   | `error` |
     | Match       | `False` |

     **Detail:** row count mismatch: feldera=0 spark=2

     ### Error

     ```
     The program failed to compile: RustError
     Rust Error: {'exit_code': 101, 'stdout': '', 'stderr': "warning: dbsp_adapters@0.266.0: not within a suitable 'git' worktree!
     warning: dbsp_adapters@0.266.0: VERGEN_GIT_SHA set to default
     warning: dbsp_adapters@0.266.0: VERGEN_GIT_DIRTY set to default
        Compiling feldera_pipe_operator_dd4fc6c2c5b9ed86 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_dd4fc6c2c5b9ed86)
        Compiling feldera_pipe_operator_cda923815e78070b v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_cda923815e78070b)
        Compiling feldera_pipe_operator_bd537152fbfe8610 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_bd537152fbfe8610)
        Compiling feldera_pipe_operator_25183f7abe968481 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_25183f7abe968481)
        Compiling feldera_pipe_operator_0269bb98f62f16e9 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_0269bb98f62f16e9)
        Compiling feldera_pipe_operator_a174b404f07c6517 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_a174b404f07c6517)
        Compiling feldera_pipe_operator_fa18e0b2b218b26b v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_fa18e0b2b218b26b)
     error[E0507]: cannot move out of `*p1` which is behind a shared reference
       --> crates/feldera_pipe_operator_cda923815e78070b/src/lib.rs:80:17
        |
     80 |                 (*p1), (
        |                 ^^^^^ move occurs because `*p1` has type `Arc<BTreeMap<i32, i32>>`, which does not implement the `Copy` trait

     error[E0507]: cannot move out of `*p1` which is behind a shared reference
       --> crates/feldera_pipe_operator_0269bb98f62f16e9/src/lib.rs:80:17
        |
     80 |                 (*p1), (
        |                 ^^^^^ move occurs because `*p1` has type `Arc<BTreeMap<i32, i32>>`, which does not implement the `Copy` trait

     error[E0507]: cannot move out of `*p1` which is behind a shared reference
       --> crates/feldera_pipe_operator_dd4fc6c2c5b9ed86/src/lib.rs:79:17
        |
     79 |                 (*p1), (
        |                 ^^^^^ move occurs because `*p1` has type `Arc<BTreeMap<i32, i32>>`, which does not implement the `Copy` trait

     For more information about this error, try `rustc --explain E0507`.
     error: could not compile `feldera_pipe_operator_cda923815e78070b` (lib) due to 1 previous error
     warning: build failed, waiting for other jobs to finish...
     error: could not compile `feldera_pipe_operator_dd4fc6c2c5b9ed86` (lib) due to 1 previous error
     error: could not compile `feldera_pipe_operator_0269bb98f62f16e9` (lib) due to 1 previous error
     "}

     ```

### Feldera SQL (translated)

     **Schema**

     ```sql
     CREATE TABLE employee_records (emp_id INT, hire_date DATE, department VARCHAR, salary INT);
     ```

     **Query**

     ```sql
     CREATE VIEW quarterly_hires_v2 AS SELECT emp_id, hire_date, DATE_TRUNC(hire_date, QUARTER) as quarter_start, DATE_TRUNC(hire_date, DAY) as day_start, department FROM employee_records WHERE department IS NOT NULL ORDER BY hire_date;
     ```

     ### Status

     | Phase | Status |
     |---|---|
     | Translation | `success` |
     | Compilation | `ok` |
     | Execution   | `error` |
     | Match       | `False` |

     **Detail:** row count mismatch: feldera=0 spark=6

     ### Error

     ```
     The program failed to compile: RustError
     Rust Error: {'exit_code': 101, 'stdout': '', 'stderr': "warning: dbsp_adapters@0.272.0: not within a suitable 'git' worktree!
     warning: dbsp_adapters@0.272.0: VERGEN_GIT_SHA set to default
     warning: dbsp_adapters@0.272.0: VERGEN_GIT_DIRTY set to default
        Compiling feldera_pipe_operator_1338c7271dcbc115 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_1338c7271dcbc115)
        Compiling feldera_pipe_operator_4a23c2f3c15642f6 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_4a23c2f3c15642f6)
        Compiling feldera_pipe_pipeline_019d178f_cc5b_7422_ab89_0bc24e29f65d_globals v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_pipeline_019d178f_cc5b_7422_ab89_0bc24e29f65d_globals)
        Compiling feldera_pipe_operator_4265b9b19b43e210 v0.1.0 (/home/ubuntu/.feldera/compiler/rust-compilation/crates/feldera_pipe_operator_4265b9b19b43e210)
     error[E0425]: cannot find function `date_trunc_quarter_DateN` in this scope
         --> crates/feldera_pipe_operator_4265b9b19b43e210/src/lib.rs:80:17
          |
       80 |                 date_trunc_quarter_DateN((*p0).1.as_ref().cloned()),
          |                 ^^^^^^^^^^^^^^^^^^^^^^^^ help: a function with a similar name exists: `date_trunc_year_DateN`
          |
         ::: /home/ubuntu/feldera/crates/sqllib/src/timestamp.rs:1997:1
          |
     1997 | some_polymorphic_function1!(date_trunc_year, Date, Date, Date);
          | -------------------------------------------------------------- similarly named function `date_trunc_year_DateN` defined here

     For more information about this error, try `rustc --explain E0425`.
     error: could not compile `feldera_pipe_operator_4265b9b19b43e210` (lib) due to 1 previous error
     warning: build failed, waiting for other jobs to finish...
     "}

     ```

--- a/crates/sqllib/src/lib.rs
+++ b/crates/sqllib/src/lib.rs
@@ -457,7 +457,21 @@ macro_rules! some_function4 {
             }
 
             #[doc(hidden)]
-            pub fn [<$func_name N__N>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: Option<$arg_type2>, arg3: Option<$arg_type3> ) -> Option<$ret_type> {
+            pub fn [<$func_name N__N>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: $arg_type2, arg3: Option<$arg_type3> ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                let arg3 = arg3?;
+                Some([<$func_name ____>](arg0, arg1, arg2, arg3))
+            }
+
+            #[doc(hidden)]
+            pub fn [<$func_name N_N_>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: Option<$arg_type2>, arg3: $arg_type3 ) -> Option<$ret_type> {
+                let arg0 = arg0?;
+                let arg2 = arg2?;
+                Some([<$func_name ____>](arg0, arg1, arg2, arg3))
+            }
+
+            #[doc(hidden)]
+            pub fn [<$func_name N_NN>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: Option<$arg_type2>, arg3: Option<$arg_type3> ) -> Option<$ret_type> {
                 let arg0 = arg0?;
                 let arg2 = arg2?;
                 let arg3 = arg3?;
@@ -465,31 +479,16 @@ macro_rules! some_function4 {
             }
 
             #[doc(hidden)]
-            pub fn [<$func_name N_N_>]( arg0: Option<$arg_type0>, arg1: Option<$arg_type1>, arg2: $arg_type2, arg3: $arg_type3 ) -> Option<$ret_type> {
+            pub fn [<$func_name NN__>]( arg0: Option<$arg_type0>, arg1: Option<$arg_type1>, arg2: $arg_type2, arg3: $arg_type3 ) -> Option<$ret_type> {
                 let arg0 = arg0?;
                 let arg1 = arg1?;
                 Some([<$func_name ____>](arg0, arg1, arg2, arg3))
             }
 
             #[doc(hidden)]
-            pub fn [<$func_name N_NN>]( arg0: Option<$arg_type0>, arg1: Option<$arg_type1>, arg2: Option<$arg_type2>, arg3: Option<$arg_type3> ) -> Option<$ret_type> {
+            pub fn [<$func_name NN_N>]( arg0: Option<$arg_type0>, arg1: Option<$arg_type1>, arg2: $arg_type2, arg3: Option<$arg_type3> ) -> Option<$ret_type> {
                 let arg0 = arg0?;
                 let arg1 = arg1?;
-                let arg2 = arg2?;
-                let arg3 = arg3?;
-                Some([<$func_name ____>](arg0, arg1, arg2, arg3))
-            }
-
-            #[doc(hidden)]
-            pub fn [<$func_name NN__>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: $arg_type2, arg3: $arg_type3 ) -> Option<$ret_type> {
-                let arg0 = arg0?;
-                Some([<$func_name ____>](arg0, arg1, arg2, arg3))
-            }
-
-            #[doc(hidden)]
-            pub fn [<$func_name NN_N>]( arg0: Option<$arg_type0>, arg1: $arg_type1, arg2: Option<$arg_type2>, arg3: Option<$arg_type3> ) -> Option<$ret_type> {
-                let arg0 = arg0?;
-                let arg2 = arg2?;
                 let arg3 = arg3?;
                 Some([<$func_name ____>](arg0, arg1, arg2, arg3))
             }

--- a/crates/sqllib/src/timestamp.rs
+++ b/crates/sqllib/src/timestamp.rs
@@ -1853,6 +1853,13 @@ pub fn date_trunc_year_Timestamp(timestamp: Timestamp) -> Timestamp {
 }
 
 #[doc(hidden)]
+pub fn date_trunc_quarter_Timestamp(timestamp: Timestamp) -> Timestamp {
+    let dt = timestamp.get_date();
+    let dt = date_trunc_quarter_Date(dt);
+    Timestamp::from_date(dt)
+}
+
+#[doc(hidden)]
 pub fn date_trunc_month_Timestamp(timestamp: Timestamp) -> Timestamp {
     let dt = timestamp.get_date();
     let dt = date_trunc_month_Date(dt);
@@ -1877,6 +1884,7 @@ some_polymorphic_function1!(date_trunc_millennium, Timestamp, Timestamp, Timesta
 some_polymorphic_function1!(date_trunc_century, Timestamp, Timestamp, Timestamp);
 some_polymorphic_function1!(date_trunc_decade, Timestamp, Timestamp, Timestamp);
 some_polymorphic_function1!(date_trunc_year, Timestamp, Timestamp, Timestamp);
+some_polymorphic_function1!(date_trunc_quarter, Timestamp, Timestamp, Timestamp);
 some_polymorphic_function1!(date_trunc_month, Timestamp, Timestamp, Timestamp);
 some_polymorphic_function1!(date_trunc_week, Timestamp, Timestamp, Timestamp);
 some_polymorphic_function1!(date_trunc_day, Timestamp, Timestamp, Timestamp);
@@ -1899,6 +1907,11 @@ pub fn timestamp_trunc_decade_Timestamp(timestamp: Timestamp) -> Timestamp {
 #[doc(hidden)]
 pub fn timestamp_trunc_year_Timestamp(timestamp: Timestamp) -> Timestamp {
     date_trunc_year_Timestamp(timestamp)
+}
+
+#[doc(hidden)]
+pub fn timestamp_trunc_quarter_Timestamp(timestamp: Timestamp) -> Timestamp {
+    date_trunc_quarter_Timestamp(timestamp)
 }
 
 #[doc(hidden)]
@@ -1986,6 +1999,20 @@ pub fn date_trunc_year_Date(date: Date) -> Date {
 }
 
 #[doc(hidden)]
+pub fn date_trunc_quarter_Date(date: Date) -> Date {
+    let date = date.to_dateTime();
+    let month = match date.month() {
+        1..=3 => 1,
+        4..=6 => 4,
+        7..=9 => 7,
+        10..=12 => 10,
+        _ => unreachable!(),
+    };
+    let rounded = NaiveDate::from_ymd_opt(date.year(), month, 1).unwrap();
+    Date::from_date(rounded)
+}
+
+#[doc(hidden)]
 pub fn date_trunc_month_Date(date: Date) -> Date {
     let date = date.to_dateTime();
     let rounded = NaiveDate::from_ymd_opt(date.year(), date.month(), 1).unwrap();
@@ -2009,6 +2036,7 @@ some_polymorphic_function1!(date_trunc_millennium, Date, Date, Date);
 some_polymorphic_function1!(date_trunc_century, Date, Date, Date);
 some_polymorphic_function1!(date_trunc_decade, Date, Date, Date);
 some_polymorphic_function1!(date_trunc_year, Date, Date, Date);
+some_polymorphic_function1!(date_trunc_quarter, Date, Date, Date);
 some_polymorphic_function1!(date_trunc_month, Date, Date, Date);
 some_polymorphic_function1!(date_trunc_week, Date, Date, Date);
 some_polymorphic_function1!(date_trunc_day, Date, Date, Date);

--- a/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/expandCasts/ExpandUnsafeCasts.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/main/java/org/dbsp/sqlCompiler/compiler/visitors/outer/expandCasts/ExpandUnsafeCasts.java
@@ -253,7 +253,7 @@ public class ExpandUnsafeCasts extends ExpressionTranslator {
         DBSPType convertedType = new DBSPTypeMap(type.getKeyType(), type.getValueType(), sourceType.mayBeNull);
         return new DBSPBinaryExpression(node,
                 convertedType, DBSPOpcode.MAP_CONVERT,
-                source, new DBSPRawTupleExpression(convertKey, convertValue));
+                source.applyClone(), new DBSPRawTupleExpression(convertKey, convertValue));
     }
 
     @Override

--- a/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression2Tests.java
+++ b/sql-to-dbsp-compiler/SQL-compiler/src/test/java/org/dbsp/sqlCompiler/compiler/sql/simple/Regression2Tests.java
@@ -823,4 +823,54 @@ public class Regression2Tests extends SqlIoTest {
                     jsonstring_as_a_t("data")."b"."y"."value" AS "y_val"
                 FROM src;""");
     }
+
+    @Test
+    public void issue5899a() {
+        this.getCCS("""
+                CREATE TABLE F(file_id INT, original_path VARCHAR, replacement VARCHAR);
+                CREATE VIEW V AS SELECT file_id, original_path,
+                OVERLAY(original_path PLACING replacement FROM 10 FOR 4) AS updated_path
+                FROM F WHERE original_path IS NOT NULL AND replacement IS NOT NULL;""");
+        boolean[] b = new boolean[] { false, true };
+        for (var a1: b)
+            for (var a2: b)
+                for (var a3: b)
+                    for (var a4: b) {
+                        var e1 = a1 ? "x" : "y";
+                        var e2 = a2 ? "x" : "y";
+                        var e3 = a3 ? "1" : "NULL";
+                        var e4 = a4 ? "1" : "NULL";
+                        String query = "CREATE TABLE T(x VARCHAR NOT NULL, y VARCHAR, i INT NOT NULL);";
+                        query += "CREATE VIEW V AS SELECT OVERLAY(" + e1 +
+                                " PLACING " + e2 + " FROM " + e3 + " FOR " + e4 + ") FROM T;";
+                        this.getCCS(query);
+                    }
+    }
+
+    @Test
+    public void issue5899b() {
+        this.getCCS("""
+                CREATE VIEW V AS
+                SELECT COUNT(*)
+                FROM (
+                  VALUES
+                    (CAST(ROW(ARRAY[MAP[1, 2, 2, 3], MAP[1, 3]]) AS ROW(b MAP<INT, INT> ARRAY))),
+                    (CAST(ROW(ARRAY[MAP[2, 3], MAP[1, 3]]) AS ROW(b MAP<INT, INT> ARRAY))),
+                    (CAST(ROW(ARRAY[MAP[2, 3, 1, 2], MAP[1, 3]]) AS ROW(b MAP<INT, INT> ARRAY)))
+                ) AS t(a)
+                GROUP BY a""");
+    }
+
+    @Test
+    public void issue5899c() {
+        var ccs = this.getCCS("""
+                CREATE TABLE E(hire_date DATE);
+                CREATE VIEW Q AS
+                SELECT DATE_TRUNC(hire_date, QUARTER) FROM E;""");
+        ccs.step("INSERT INTO E VALUES (DATE '2020-02-01') , (DATE '2020-05-01')", """
+                 trunc | weight
+                ----------------
+                 2020-01-01 | 1
+                 2020-04-01 | 1""");
+    }
 }

